components/file_system/src/condvar.rs

```diff
@@ -0,0 +1,303 @@
+unsafe impl Sync for Condvar {}
+
+impl Condvar {
+    pub fn new() -> Condvar {
+        Condvar {
+            head: Cell::new(None),
+            tail: Cell::new(None),
+        }
+    }
+
+    #[inline]
+    fn enqueue(&self, raw_node: &mut CondvarNode) {
+        let node = unsafe { Some(NonNull::new_unchecked(raw_node)) };
+        raw_node.set_prev(self.tail.get());
+        if let Some(tail) = self.tail.get() {
+            unsafe {
+                tail.as_ref().set_next(node);
+            }
+        } else {
+            self.head.set(node);
+        }
+        self.tail.set(node);
+    }
+
+    #[inline]
+    fn dequeue(&self, raw_node: &mut CondvarNode) {
+        let prev = raw_node.get_prev();
+        let next = raw_node.get_next();
+        if let Some(prev) = prev {
+            unsafe {
+                prev.as_ref().set_next(next);
+            }
+        } else {
+            assert!(self.head.get().unwrap().as_ptr() == raw_node);
+            self.head.set(next);
+        }
+        if let Some(next) = next {
+            unsafe {
+                next.as_ref().set_prev(prev);
+            }
+        } else {
+            assert!(self.tail.get().unwrap().as_ptr() == raw_node);
+            self.tail.set(prev);
+        }
+    }
+
+    /// Notifies all waiters in queue as till now.
+    pub fn notify_all(&self) {
+        let mut ptr = self.head.get();
+        while let Some(inner) = ptr {
+            unsafe {
+                let node = inner.as_ref();
+                node.notify();
+                ptr = node.get_next();
+            }
+        }
+    }
+
+    /// Synchronously waits on this condition variable for a notification,
+    /// timing out after a specified duration.
+    pub fn wait_timeout<'a, T>(
+        &self,
+        mut guard: MutexGuard<'a, T>,
+        timeout: Duration,
+    ) -> (MutexGuard<'a, T>, bool) {
+        // mutable just to indulge NonNull
+        let mut node = CondvarNode::new_sync();
+        self.enqueue(&mut node);
+        // alternative: std::thread::park_timeout suffers from spurious wake
+        let res = match node {
+            CondvarNode::Sync(ref condv, _) => condv.wait_for(&mut guard, timeout),
+            _ => unreachable!(),
+        };
+        self.dequeue(&mut node);
+        (guard, res.timed_out())
+    }
+
+    /// Asynchronously waits on this condition variable for a notification,
+    /// timing out after a specified duration. Need to pass in additional
+    /// reference of the original mutex to regain lock after wakeup.
+    pub async fn async_wait_timeout<'a, 'b, T>(
+        &self,
+        mu: &'a Mutex<T>,
+        guard: MutexGuard<'b, T>,
+        timeout: Duration,
+    ) -> (MutexGuard<'a, T>, bool) {
+        let mut node = CondvarNode::new_async();
+        self.enqueue(&mut node);
+        std::mem::drop(guard);
+        let timed_out = {
+            let f = match node {
+                CondvarNode::Async(ref sem, _) => sem.acquire().fuse(),
```

Comment:
use tokio::sync::Notify?

Reply:
Seems the only difference is whether the notifications can accumulate, not sure the performance-wise which is the winner. I'll leave a todo here for future reference.